src/resolvers/deck/createDeck.ts

```diff
@@ -28,3 +35,190 @@ export const createDeck = mutationWithClientMutationId({
+      })
+    }
+
+    // Create deck
+    const newDeck = await DeckModel.create({
+      title: `${publishedDeck?.title} - new`,
+      description: publishedDeck?.description,
+      ownerId: user?._id,
+      slug: '',
+      published: false,
+      configuration: defaultDeckConfig,
+    })
+
+    // Create models/fields & templates
+    for await (const model of models) {
+      // Create model
+      const newModel = await ModelModel.create({
+        name: `${model.model?.name} - copy`,
+        ownerId: user?._id,
+      })
+      helperMap.updateModelIndex.push({
+        oldId: model.model._id,
+        newId: newModel._id,
+      })
+
+      // Create fields
+      for await (const field of model.fields) {
+        const currentField = await FieldModel.create({
+          name: field.name,
+          modelId: newModel._id,
+        })
+        helperMap.updateFieldIndex.push({
+          oldId: field._id,
+          newId: currentField._id,
+        })
+      }
+
+      // Create templates
+      for await (const template of model.templates) {
+        const frontSideWithUpdatedEntities = updateEntitiesIds(
+          helperMap.updateFieldIndex,
+          template.frontSide
+        )
+        const backSideWithUpdatedEntities = updateEntitiesIds(
+          helperMap.updateFieldIndex,
+          template.backSide
+        )
+
+        await TemplateModel.create({
+          name: template.name,
+          modelId: newModel._id,
+          ownerId: user?._id,
+          frontSide: frontSideWithUpdatedEntities,
+          backSide: backSideWithUpdatedEntities,
+        })
+      }
+    }
+
+    // Create notes
+    for await (const note of deckNotes) {
+      const currentNoteModelId = helperMap.updateModelIndex.find(
+        (v) => `${v.oldId}` === `${note.modelId}`
+      )?.newId
+
+      const fieldValues = note.values.map((noteValue) => {
+        const newFieldId = helperMap.updateFieldIndex.find(
+          (v) => `${v.oldId}` === `${noteValue.fieldId}`
+        )?.newId
+
+        return {
+          data: removeContentStateDocumentIds(noteValue.data),
+          fieldId: newFieldId,
+        }
+      })
+
+      const newNote = await NoteModel.create({
+        modelId: currentNoteModelId,
+        deckId: newDeck._id,
+        ownerId: user!._id,
+        values: fieldValues,
+        flashCards: [],
+      })
+
+      // Create flashcards
+      const modelTemplates = await TemplateModel.find({
+        modelId: currentNoteModelId,
+      })
+
+      newNote.set(
+        'flashCards',
+        modelTemplates.map(({ _id: templateId }) => ({
+          templateId,
+          noteId: note._id,
+        }))
+      )
+
+      await newNote.save()
+    }
+    return { deck: newDeck }
+  },
+})
+
+const removeContentStateDocumentIds = (
+  contentState: ContentStateInput | undefined
+): ContentStateInput | undefined => {
+  if (!contentState) return undefined
+  const cleanBlocks = contentState.blocks.map((b) => {
+    return {
```

Comment:
can you use spread here? I don't get what this function is doing besides doing a shallow clone of the blocks

Reply:
The idea here was to remove _ids on the Note Document. It should remove these _ids

![image](https://user-images.githubusercontent.com/14936466/144672823-1105b225-5ab3-4535-a4ed-1e1c877051f3.png)

This is an example of one `contentState` object

```json
{"blocks":[{"key":"d0kvm","type":"unstyled","text":"17","inlineStyleRanges":[],"entityRanges":[],"depth":0,"_id":"61a56c6f5d501b90adea12bc"}],"_id":"61a56c6f5d501b90adea12bb"}
```

Reply:
If I try to do something like:
```typescript
const removeContentStateDocumentIds = (
  contentState: ContentStateInput | undefined
): ContentStateInput | undefined => {
  if (!contentState) return undefined
  return {
    blocks: contentState.blocks.map((b) => {
      return {
        ...b,
        _id: undefined,
      }
    }),
  }
}
```


Then there is a "problem" related with the mongoDB object type attributes:
```
[GraphQL error]: Message Note validation failed: values.0.data.blocks: Cast to Array failed for value "[
  {
    __parentArray: [ [Object] ],
    __index: 0,
    '$__parent': { blocks: [Array], _id: new ObjectId("61aa8cb79197f32b6da87d5f") },
    '$__': InternalCache {
      activePaths: [StateMachine],
      ownerDocument: [Object],
      fullPath: 'values.data.blocks',
      emitter: [EventEmitter],
      '$options': [Object],
      strictMode: true,
      selected: undefined,
      _id: new ObjectId("61aa8cb79197f32b6da87d60"),
      session: null
    },
    '$isNew': false,
    '$op': null,
    _doc: {
      key: 'cihvr',
      type: 'unstyled',
      text: 'Value filed 1',
      inlineStyleRanges: [],
      entityRanges: [],
      depth: 0,
      _id: new ObjectId("61aa8cb79197f32b6da87d60")
    },
    '$init': true,
    _id: undefined,
    [Symbol(mongoose:documentArrayParent)]: { blocks: [Array], _id: new ObjectId("61aa8cb79197f32b6da87d5f") }
  }
]"
```

It related to: [StackOverflow: Object.keys() returns unexpected keys on MongoDB object from collection](https://stackoverflow.com/questions/44833736/object-keys-returns-unexpected-keys-on-mongodb-object-from-collection/44833913)

If I spread I get these keys:
```
[ '__parentArray',
  '__parent',
  '__index',
  '$__',
  'isNew',
  'errors',
  '_doc',
  '$init' ]
```

Reply:
I'll test something, but if it doesn't work out I think this is good enough

Reply:
alright, so you should be able to convert the block into a JSON object using the `.toJSON()` method, but you'll need to update the definition of the `ContentStateDocument` interface, like so

```patch
diff --git a/src/mongo/ContentState.ts b/src/mongo/ContentState.ts
index b2626b2..7b805af 100644
--- a/src/mongo/ContentState.ts
+++ b/src/mongo/ContentState.ts
@@ -12,26 +12,32 @@ export type EntityMapValue =
     }
   | { type: 'LINK'; mutability: DraftEntityMutability; data: { url: string } }
 
-export interface ContentState {
-  blocks: {
-    key: string
-    type: string
-    text: string
-    depth: number
-    inlineStyleRanges: RawDraftInlineStyleRange[]
-    entityRanges: {
-      key: number
-      length: number
-      offset: number
-    }[]
-    data: BlockData
+interface Block {
+  key: string
+  type: string
+  text: string
+  depth: number
+  inlineStyleRanges: RawDraftInlineStyleRange[]
+  entityRanges: {
+    key: number
+    length: number
+    offset: number
   }[]
+  data: BlockData
+}
+
+export interface ContentState {
+  blocks: Block[]
   entityMap: {
     [key: string]: EntityMapValue
   }
 }
 
-export interface ContentStateDocument extends ContentState, Document {}
+interface BlockDocument extends Block, Document {}
+
+export interface ContentStateDocument extends ContentState, Document {
+  blocks: BlockDocument[]
+}
 
 export const ContentStateSchema = new Schema<ContentStateDocument>({
   blocks: [
diff --git a/src/resolvers/deck/createDeck.ts b/src/resolvers/deck/createDeck.ts
index e605a4f..3e53460 100644
--- a/src/resolvers/deck/createDeck.ts
+++ b/src/resolvers/deck/createDeck.ts
@@ -8,7 +8,10 @@ import {
   NoteModel,
   TemplateModel,
 } from '../../mongo'
-import type { ContentState } from '../../mongo/ContentState'
+import type {
+  ContentState,
+  ContentStateDocument,
+} from '../../mongo/ContentState'
 import { defaultDeckConfig } from '../../mongo/Deck'
 import { DeckType } from './types'
 
@@ -185,19 +188,14 @@ export const importDeck = mutationWithClientMutationId({
 })
 
 const removeContentStateDocumentIds = (
-  contentState: ContentStateInput | undefined
+  contentState: ContentStateDocument | undefined
 ): ContentStateInput | undefined => {
   if (!contentState) return undefined
-  const cleanBlocks = contentState.blocks.map((b) => {
-    return {
-      key: b.key,
-      type: b.type,
-      text: b.text,
-      inlineStyleRanges: b.inlineStyleRanges,
-      entityRanges: b.entityRanges,
-      depth: b.depth,
-      data: b.data,
-    }
+
+  const cleanBlocks = contentState.blocks.map((blockDocument) => {
+    const { _id, ...block } = blockDocument.toJSON()
+
+    return block
   })
 
   return {
```

Reply:
Thanks!